src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,241 @@
+        CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-        cart.setTotal(cart.getTotal() - (products.size() * product.getProductPrice()));
-
-        for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-            productTotalRemove = product.getProductPrice() + productTotalRemove;
+        if (cartEntry == null) {
+            throw new EntityNotFoundException("CartEntry not found in the Database!");
+        } else {
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
         }
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
 
-        cart.setProductList(productList);
+        double total = cart.getTotal() - (initialQuantity * productDTO.getPrice()) + (quantity * productDTO.getPrice());
 
-        cartRepository.save(cart);
-
-        return CartMapper.cartEntityToDto(cart);
+        cart.setTotal(total);
 
+        cartRepository.save(cart);
 
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
-    public CartDTO removeProductFromCart(String email, String productName, int quantity)
+    public CartDTO removeProductFromCart(String email, String productName)
             throws InvalidInputDataException, EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
+        if (productDTO.getStock() == 0) {
+            throw new InvalidInputDataException("We are sorry, but currently: " + productName + " is out of order!");
+        }
+
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
 
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
+        cartEntryRepository.delete(cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().get());
```

Comment:
it would be enough to remove the found cartEntry from the cartEntries list(without the need to call any cartEntryRepository method). you will need to set orphanRemoval=true on the OneToMany association.
the cartRepository.save will be enough.

Reply:
I'm really astonished about your technical skills! 👏 
Done!

Reply:
thank you! don't be so astonished. I've played a little bit yesterday just to be sure that I'm not saying something wrong. :)